python/cudf_polars/cudf_polars/experimental/rapidsmpf/nodes.py

```diff
@@ -0,0 +1,239 @@
+    ch_out
+        The output channel.
+    chs_in
+        The input channels.
+    """
+    # TODO: Use multiple streams
+    n_children = len(chs_in)
+    async with shutdown_on_error(ctx, *chs_in, ch_out):
+        # Collect broadcasted partitions first
+        broadcasted_data = {}
+        for i in range(len(broadcasted)):
+            if (msg := await chs_in[i].recv(ctx)) is not None:
+                broadcasted_data[i] = DataFrame.from_table(
+                    TableChunk.from_message(msg).table_view(),
+                    list(ir.children[i].schema.keys()),
+                    list(ir.children[i].schema.values()),
+                )
+
+        seq_num = 0
+        while True:
+            input_dfs: list[DataFrame] = []
+            for i, ch_in in enumerate(chs_in):
+                if (_df := broadcasted_data.get(i)) is not None:
+                    input_dfs.append(_df)
+                elif (msg := await ch_in.recv(ctx)) is not None:
+                    input_dfs.append(
+                        DataFrame.from_table(
+                            TableChunk.from_message(msg).table_view(),
+                            list(ir.children[i].schema.keys()),
+                            list(ir.children[i].schema.values()),
+                        )
+                    )
+
+            if len(input_dfs) == n_children:
+                # Evaluate the IR node
+                df: DataFrame = ir.do_evaluate(*ir._non_child_args, *input_dfs)
+
+                # Return the output chunk
+                chunk = TableChunk.from_pylibcudf_table(
+                    seq_num, df.table, DEFAULT_STREAM
+                )
+                await ch_out.send(ctx, Message(chunk))
+                seq_num += 1
+            elif len(broadcasted) == len(input_dfs):
+                break  # All done
+            else:
+                raise ValueError(
+                    "Number of input chunks does not match the child count."
+                )
+            if len(broadcasted) == n_children:
+                break  # All done
+        await ch_out.drain(ctx)
+
+
+@define_py_node()
+async def multicast_node(
+    ctx: Context,
+    ch_in: Channel[TableChunk],
+    *chs_out: Channel[TableChunk],
+) -> None:
+    """
+    Fan-out node for rapidsmpf.
+
+    Parameters
+    ----------
+    ctx
+        The context.
+    ch_in
+        The input channel.
+    chs_out
+        The output channels.
+    """
+    # TODO: Use multiple streams
+    async with shutdown_on_error(ctx, ch_in, *chs_out):
+        # Collect all chunks from input channel
+        chunks: list[TableChunk] = []
+        seq_num = 0
+        while (msg := await ch_in.recv(ctx)) is not None:
+            chunks.append(TableChunk.from_message(msg))
+            print(f"Received chunk {seq_num} from input channel", flush=True)
+            seq_num += 1
+
+        # Send chunks to all output channels
+        for i, ch_out in enumerate(chs_out):
```

Comment:
@madsbk @wence- - This PR is still very rough, but quick question: Are we allowed to push the same data to multiple channels like this? Any pointers on this?

In cudf-polars, an IR node may be consumed by multiple other IR nodes, so we get the wrong results in some cases with a single output channel.